File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>

using namespace std;

struct DATA {
    string file;
    string head;
    int num;
};

// 정렬 기준
bool cmp(const DATA &a, const DATA &b) {
    if (a.head == b.head) {
        return a.num < b.num;
    }
    return a.head < b.head;
}

vector<string> solution(vector<string> files) {
    vector<string> answer;
    vector<DATA> v;

    for (const string& file : files) {
        string head = "";
        string numStr = "";
        int i = 0;

        // 1. HEAD 추출
        while (i < file.length() && !isdigit(file[i])) {
            head += file[i++];
        }

        // 2. NUMBER 추출 (최대 5글자)
        while (i < file.length() && isdigit(file[i]) && numStr.length() < 5) {
            numStr += file[i++];
        }

        // HEAD는 비교를 위해 소문자로 통일
        string lowerHead = head;
        transform(lowerHead.begin(), lowerHead.end(), lowerHead.begin(), ::tolower);

        v.push_back({file, lowerHead, stoi(numStr)});
    }

    // ⭐ 핵심: std::sort 대신 std::stable_sort 사용
    stable_sort(v.begin(), v.end(), cmp);

    for (auto& item : v) {
        answer.push_back(item.file);
    }
    return answer;
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.io.*;
import java.util.*;
import java.lang.*;

class Solution {
    public String[] solution(String[] files) {
        String[] answer = {};
        
        Arrays.sort(files, (f1, f2) -> {
            String[] parts1 = splitFileName(f1);
            String[] parts2 = splitFileName(f2);
            
            int headCmp = parts1[0].equalsIgnoreCase(parts2[0]) ? 0 : parts1[0].compareToIgnoreCase(parts2[0]);
            if(headCmp != 0)
                return headCmp;
            int num1 = Integer.parseInt(parts1[1]);
            int num2 = Integer.parseInt(parts2[1]);
            
            return Integer.compare(num1, num2);
        });
        
        answer = files;
        
        return answer;
    }
    
    public String[] splitFileName(String file){
        int idx = 0;
        String head = "";
        String num = "";
        for(int i = idx; i < file.length(); i++){
            char c = file.charAt(i);
            if(c >= '0' && c <= '9')
                break;
            head += c;
            idx += 1;
        }
        
        for(int i = idx; i < file.length(); i++){
            char c = file.charAt(i);
            if(!(c >= '0' && c <= '9') || num.length() >= 5)
                break;
            num += c;
            idx += 1;
        }
        
        return new String[]{head, num};
    }
}